src/libsync/vfs/cfapi/cfapiwrapper.cpp

```diff
@@ -180,24 +206,27 @@ void CALLBACK cfApiFetchDataCallback(const CF_CALLBACK_INFO *callbackInfo, const
         alignAndSendData(receivedData);
     });
 
-    QObject::connect(vfs, &OCC::VfsCfApi::hydrationRequestFinished, &loop, [&](const QString &id, int s) {
+    QObject::connect(vfs, &OCC::VfsCfApi::hydrationRequestFinished, &loop, [&](const QString &id) {
         qDebug(lcCfApiWrapper) << "Hydration finished for request" << id;
         if (requestId == id) {
-            const auto status = static_cast<OCC::HydrationJob::Status>(s);
-            qCInfo(lcCfApiWrapper) << "Hydration done for" << path << requestId << status;
-            if (status != OCC::HydrationJob::Success) {
-                sendTransferError();
-            }
+            socket.close();
+            signalSocket.close();
             loop.quit();
         }
     });
 
     loop.exec();
 
-    if (!protrudingData.isEmpty()) {
+    if (!hydrationRequestCancelled && !protrudingData.isEmpty()) {
         qDebug(lcCfApiWrapper) << "Send remaining protruding data. Size:" << protrudingData.size();
         sendTransferInfo(protrudingData, dataOffset);
     }
+
+    const auto hydrationJobResult = static_cast<OCC::HydrationJob::Status>(vfs->finalizeHydrationJob(requestId));
```

Comment:
to avoid data races within `finalyzeHydrationJob` maybe use a similar way to call it like
```
const auto invokeResult = 
  QMetaObject::invokeMethod(vfs, [=] { vfs->cancelHydration(requestId, path); },
                            Qt::QueuedConnection);
```

Reply:
done